### PR TITLE
Added "Receive From Socar" generator (ext) node

### PR DESCRIPTION
**Pull request by aachman98**
## Addressed problem description

Addition of a generator node to receive data input from Sorcar node output. This is a dedicated node which presents a one-sided connection between Sverchok & Sorcar where mesh data is transferred **from** the latter.

Node Name: **Receive From Sorcar**
Node Category: **Generators -> Generators Extended**
*Class Name: SvReceiveFromSorcarNode*

Screenshot from 2020-05-25 19-35-15

image

### Usage

The node comes with 6 outputs: Vertices, Edges, Faces, VerticesMask, EdgesMask, FacesMask
Each of the socket corresponds to the components of the mesh data being stored in the node properties (namely "verts", "edges", "faces") and the their selection mask.

#### Default

By default, a simple Sorcar logo mesh is contained. The node behaves as a single-mesh-output node unless paired with Send to Sverchok node in the Sorcar nodetree.

image

#### Referenced

This node can be referenced inside Sorcar, and the mesh data of the object (passed as input) is transferred to this node. The node properties hold the data (unless modified externally) allowing users to remove/modify Sorcar nodetree or the object in the scene without losing the geometry.

Screenshot from 2020-05-25 19-41-06
Screenshot from 2020-05-25 19-41-12

### Parameter updates

This node will update in the same way as expected from all the other Sverchok nodes. Since the mesh data and component masks are stored internally, re-evaluating Sorcar nodetree is not necessary.
In case of any update of parameters in Sorcar, the new mesh will be set in the properties **once** when the node gets executed, and will act as the new starting point. No further evaluations are required.

In the following examples, I first modify the Sverchok nodetree (added "Unsubdivide" node) and then modify a Sorcar node parameter ("Offset" in Poke node) to demonstrate the seamless non-destructive workflow.

Screenshot from 2020-05-25 19-41-59

Screenshot from 2020-05-25 19-42-21

Video demo: https://www.youtube.com/watch?v=pTIiMj5GpyI

## Preflight checklist

- [x] Code changes complete.
- [x] Code documentation complete.
- [x] Documentation for users complete (or not required, if user never sees these changes).
- [x] Manual testing done. 
- [ ] Unit-tests implemented.
- [x] Ready for merge.

### CODE REVIEW

File: nodes/generators_extended/receive_from_sorcar.py [hunk 1]
```diff
@@ -0,0 +1,608 @@
...
+
+    def sv_init(self, context):
+        self.outputs.new('SvVerticesSocket', "Vertices")
+        self.outputs.new('SvStringsSocket', "Edges")
+        self.outputs.new('SvStringsSocket', "Polygons")
+
+
+    def process(self):
+        if self.outputs['Vertices'].is_linked:
+            self.outputs['Vertices'].sv_set(eval(self.verts))
```
Comment (zeffii) [inline nodes/generators_extended/receive_from_sorcar.py:594]: maybe more appropriate for literal data
```python
import ast

...
self.outputs['Vertices'].sv_set(ast.literal_eval(self.verts))
```

**Comment by aachman98** [comment]
I was a bit skeptical about the default mesh data (takes over 500 lines of code). If required, I'll revert back to a simple cube data (8 v, 12 e, 6 f)

**Comment by zeffii** [comment]
we can provide a geometry module in `sverchok.utils.modules.sorcar_default_geometry`  don't worry about it.

**Comment by nortikin** [comment]
COOL!

**Comment by Durman** [comment]
I expect that when a new mesh will be set to the node properties Sverchok node tree will be evaluated three times in a row.  What is not very cool.

**Comment by aachman98** [comment]
I’ll remove the “update” for verts and edges property. The props are set in this very order, so the update will be called one time only, AFTER the data has been set.

**Comment by zeffii** [comment]
there's a lot of ways to implement this kind of node, usually only a steady stream of users will show us what they really need

**Comment by Durman** [comment]
I would suggest to remove update function from all properties and add method to the node, like this:

```Python
def set_mesh(self, verts, edges, faces):
    self.verts = verts,
    self.edges = edges
    self.faces = faces
    updateNode(self, None)
```

It will make setting mesh process more clear.

**Comment by aachman98** [comment]
Nice and clean! Working on it..

**Comment by zeffii** [comment]
here's that logo compressed into `2D`.
```python
def make_sorcar_logo():
    """ makes sorcar 3d"""
    a = 0.4233
    b = 0.7778
    c = 0.6878
    d = 0.3333
    e = 1.0
    f = 0.9100
    g = 0.8678
    h = 0.2433
    i = 0.6456
    j = 0.5556
    k = 0.4656
    l = 0.2011
    m = 0.1111
    n = 0.0211

    verts_2d = [[
        (-a, -b), (-c, -b), (-b, -c), (-b, -d), (-e, -d), (-e, -f), (-f, -e), (-d, -e),
        (-d, -g), (-h, -b), (+h, -b), (+d, -g), (+d, -e), (+f, -e), (+e, -f), (+e, -d),
        (+b, -d), (+b, -c), (+c, -b), (+a, -b), (+d, -c), (+d, -i), (+a, -j), (+k, -j),
        (+j, -k), (+j, -l), (+k, -m), (+a, -m), (+d, -n), (+d, +n), (+h, +m), (-h, +m),
        (-d, +l), (-d, +k), (-h, +j), (+h, +j), (+d, +k), (+d, +a), (+a, +d), (+k, +d),
        (+j, +a), (+j, +k), (+k, +j), (+a, +j), (+d, +i), (+d, +c), (+a, +b), (+c, +b),
        (+b, +c), (+b, +d), (+e, +d), (+e, +f), (+f, +e), (+d, +e), (+d, +g), (+h, +b),
        (-h, +b), (-d, +g), (-d, +e), (-f, +e), (-e, +f), (-e, +d), (-b, +d), (-b, +c),
        (-c, +b), (-a, +b), (-d, +c), (-d, +i), (-a, +j), (-k, +j), (-j, +k), (-j, +l),
        (-k, +m), (-a, +m), (-d, +n), (-d, -n), (-h, -m), (+h, -m), (+d, -l), (+d, -k),
        (+h, -j), (-h, -j), (-d, -k), (-d, -a), (-a, -d), (-k, -d), (-j, -a), (-j, -k),
        (-k, -j), (-a, -j), (-d, -i), (-d, -c)
    ]]
    return verts_2d

```

**Comment by nortikin** [comment]
can we merge?

**Comment by aachman98** [comment]
I haven’t updated the code yet. Will push the new method with compressed mesh by tomorrow. Also, I was thinking of adding 3 new component mask outputs so that we can retain the selection.

**Comment by aachman98** [comment]
image

File: nodes/generators_extended/receive_from_sorcar.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
+import bpy
+import ast
+from bpy.props import StringProperty
+
+from sverchok.node_tree import SverchCustomTreeNode
+from sverchok.data_structure import updateNode
+from sverchok.utils.modules.geom_sorcar import make_sorcar_logo
+
+class SvReceiveFromSorcarNode(bpy.types.Node, SverchCustomTreeNode):
+    ''' Receive Mesh Data From Sorcar '''
```
Comment (zeffii) [inline nodes/generators_extended/receive_from_sorcar.py:10]: we have a useful scheme for this , and it makes tooltips for out menus automatically

```
    """
    Triggers: sorcar sc receive
    Tooltip: Receive Mesh Data From Sorcar
    
    A short description for reader of node code <---not mandatory, but you could include links of interest.
    """
```

File: nodes/generators_extended/receive_from_sorcar.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+    sc_v, sc_e, sc_f = make_sorcar_logo()
+    sv_v_mask = [[True]*len(sc_v)]
+    sv_e_mask = [[True]*len(sc_e)]
+    sv_f_mask = [[True]*len(sc_f)]
+
+
+    verts: StringProperty(name='Vertices',
+                         default=repr([sc_v]),
+                         options={'ANIMATABLE'})
```
Comment (zeffii) [inline nodes/generators_extended/receive_from_sorcar.py:24]: animatable is redundant. this is the default state of properties.

File: nodes/generators_extended/receive_from_sorcar.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        if self.outputs['PolygonsMask'].is_linked:
+            self.outputs['PolygonsMask'].sv_set(ast.literal_eval(self.faces_mask))
+
+
+def register():
+    bpy.utils.register_class(SvReceiveFromSorcarNode)
+
+
+def unregister():
+    bpy.utils.unregister_class(SvReceiveFromSorcarNode)
```
Comment (zeffii) [inline nodes/generators_extended/receive_from_sorcar.py:86]: add a new empty line

File: utils/modules/geom_sorcar.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        (-c, +b), (-a, +b), (-d, +c), (-d, +i), (-a, +j), (-k, +j), (-j, +k), (-j, +l),
+        (-k, +m), (-a, +m), (-d, +n), (-d, -n), (-h, -m), (+h, -m), (+d, -l), (+d, -k),
+        (+h, -j), (-h, -j), (-d, -k), (-d, -a), (-a, -d), (-k, -d), (-j, -a), (-j, -k),
+        (-k, -j), (-a, -j), (-d, -i), (-d, -c)
+    ]
+
+    verts_3d = [(v[0], v[1], 0.05) for v in verts_2d]
+    face = list(range(len(verts_3d)))
+    verts, edges, faces, _ = solidify(verts_3d, [face], [0.1])
+    return verts, edges, faces
```
Comment (zeffii) [inline utils/modules/geom_sorcar.py:39]: add a new empty line

Comment (zeffii) [reply utils/modules/geom_sorcar.py:39]: other than these things, it seems OK. i'm sure our adventurous users will be happy to have this kind of convenience. glad we can collaborate in this way.

File: nodes/generators_extended/receive_from_sorcar.py (hunk 2 above)
Comment (aachman98) [reply nodes/generators_extended/receive_from_sorcar.py:10]: Thanks!
I don't have any links of interest (maybe except for this PR as it has some info on the top), so I'll create a separate small description and add some comments in the code.

File: nodes/generators_extended/receive_from_sorcar.py (hunk 3 above)
Comment (aachman98) [reply nodes/generators_extended/receive_from_sorcar.py:24]: I copied the code from `generators -> sphere.py` when I started, so I thought this was the standard way Sverchok node properties were made. It also has a one-line description.

File: utils/modules/geom_sorcar.py (hunk 5 above)
Comment (aachman98) [reply utils/modules/geom_sorcar.py:39]: Thank you for all the effort!

File: nodes/generators_extended/receive_from_sorcar.py (hunk 3 above)
Comment (zeffii) [reply nodes/generators_extended/receive_from_sorcar.py]: yep, but it's not so useful to restate the default. i'm not going to retrospectively remove them, but whenever a node gets added or updated i try to nip that in the bud.

**Comment by enzyme69** [comment]
Screen Shot 2020-06-10 at 7 52 55 am

Missing "Send to Sorcar" node in SV

Screen Shot 2020-06-10 at 7 54 14 am

And also getting error when sending to SV from SC.

```
Error: Traceback (most recent call last):
  File "/Users/jimmygunawan/Library/Application Support/Blender/2.83/scripts/addons/Sorcar-master/operators/ScExecuteNode.py", line 19, in execute
    curr_tree.execute_node()
  File "/Users/jimmygunawan/Library/Application Support/Blender/2.83/scripts/addons/Sorcar-master/tree/ScNodeTree.py", line 58, in execute_node
    if (not n.execute()):
  File "/Users/jimmygunawan/Library/Application Support/Blender/2.83/scripts/addons/Sorcar-master/nodes/_base/node_base.py", line 56, in execute
    self.functionality()
  File "/Users/jimmygunawan/Library/Application Support/Blender/2.83/scripts/addons/Sorcar-master/nodes/utilities/ScSendToSverchok.py", line 46, in functionality
    self.prop_tree.nodes[self.prop_node].set_mesh(repr(v), repr(e), repr(f))
TypeError: set_mesh() missing 3 required positional arguments: 'verts_mask', 'edges_mask', and 'faces_mask'
```

**Comment by enzyme69** [comment]
However SV node remote control confirmed working with SC...

Screen Shot 2020-06-10 at 7 59 20 am

**Comment by aachman98** [comment]
> And also getting error when sending to SV from SC.

Sorry, issue from my side. I modified the "Receive From Sorcar" node in the last few commits which allows you to get selection mask as well, but the code for Sorcar node is not yet updated.
Issue created here: https://github.com/aachman98/Sorcar/issues/152
The sverchok node has no issues.


> Missing "Send to Sorcar" node in SV

Testing for bugs currently, as selection mask inputs have been added to that node as well. Will be creating a separate PR for it.

**Comment by hanoixan** [comment]
@aachman98 Did a separate PR ever get created to add the "Send to Sorcar" node? As of now, it appears to still be missing.

**Comment by nortikin** [comment]
> @aachman98 Did a separate PR ever get created to add the "Send to Sorcar" node? As of now, it appears to still be missing.

ATM `node remote control` to manipulate some values on sorcar. Or use viewer bmesh to create object and use it on any other node system

**Pull request merged**